File: src/mcsstrategy.cpp
#include "mcsstrategy.hpp"

std::ostream& operator<<(std::ostream& os, const McsStrategy& mcs_strategy) {
    if (mcs_strategy == McsStrategy::Always) { os <<  "always"; }
    else if (mcs_strategy == McsStrategy::Rescue) { os << "rescue"; }
    else if (mcs_strategy == McsStrategy::FirstStrobe) { os << "first strobe"; }
    else if (mcs_strategy == McsStrategy::Off) { os << "off"; }

    return os;
}

std::ostream& operator<<(std::ostream& os, const SearchStrategy& search_strategy) {
    if (search_strategy == SearchStrategy::First) { os <<  "first only"; }
    else if (search_strategy == SearchStrategy::Second) { os << "second only"; }
    else if (search_strategy == SearchStrategy::Rescue) { os << "second as rescue"; }

    return os;
}

File: src/mcsstrategy.rs
use std::fmt::{Display, Formatter};

#[derive(Default, Debug, PartialEq, Eq, Copy, Clone, clap::ValueEnum)]
pub enum McsStrategy {
    // For each strobemer, do a full lookup. If that did not generate a hit,
    // try a partial lookup.
    #[default]
    Always,

    // For each strobemer, do a full lookup. If after processing the entire
    // query, no hits were generated, do partial lookups of each strobemer.
    Rescue,

    // Do full lookups only.
    Off,

    // Do partial lookups only, that is, use only the first strobe.
    FirstStrobe,
}

impl Display for McsStrategy {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            McsStrategy::Always => write!(f, "Always"),
            McsStrategy::Rescue => write!(f, "Rescue"),
            McsStrategy::Off => write!(f, "Off"),
            McsStrategy::FirstStrobe => write!(f, "FirstStrobe"),
        }
    }
}
